provisioner/node_pools.go

```diff
@@ -247,21 +252,36 @@ func (p *AWSNodePoolProvisioner) prepareUserData(templateCtx *templateContext, n
 
 	cloudInitPath := path.Join(nodePoolProfilesPath, cloudInitFileName)
 	if _, err := os.Stat(cloudInitPath); err == nil {
-		return p.prepareCloudInit(templateCtx, cloudInitPath)
+		return p.prepareCloudInit(nodePoolProfilesPath, templateCtx, cloudInitPath)
 	}
 
 	return "", fmt.Errorf("no userdata file at '%s' nor '%s' found", clcPath, cloudInitPath)
 }
 
 // prepareCloudInit prepares the user data by rendering the golang template.
 // A EC2 UserData ready base64 string will be returned.
-func (p *AWSNodePoolProvisioner) prepareCloudInit(templateCtx *templateContext, cloudInitPath string) (string, error) {
+func (p *AWSNodePoolProvisioner) prepareCloudInit(nodePoolProfilesPath string, templateCtx *templateContext, cloudInitPath string) (string, error) {
+	s3Path, err := p.renderUploadPKI(templateCtx, nodePoolProfilesPath)
+	if err != nil {
+		return "", err
+	}
+	templateCtx.s3PKIPath = s3Path
+	p.logger.Debugf("Uploaded PKI to %s", s3Path)
 	rendered, err := renderTemplate(templateCtx, cloudInitPath)
 	if err != nil {
 		return "", err
 	}
-
-	return base64.StdEncoding.EncodeToString([]byte(rendered)), nil
+	var gzipBuffer bytes.Buffer
+	writer := gzip.NewWriter(&gzipBuffer)
+	_, err = writer.Write([]byte(rendered))
+	if err != nil {
+		return "", err
+	}
+	err = writer.Close()
```

Comment:
too late, please use defer after line 275

Reply:
In this case I would  argue it is not because we don't need to close the writer because there is no risk of leaving dangling open file pointer. OTOH if we don't check the the return value of `close()` we might end up with a what appears to be a successful write but has in fact failed.

Reply:
Looks like `close()` can be called multiple times. So will add it in both places.